sql/src/main/java/io/crate/execution/dml/TransportShardAction.java

```diff
@@ -182,27 +187,56 @@ public void kill(@Nullable Throwable t) {
+                                                                                   Function<Exception, T> onMappingUpdateError,
+                                                                                   ActionListener<T> listener) {
+        T result;
+        try {
+            result = execute.get();
+        } catch (Exception e) {
+            listener.onFailure(e);
+            return;
+        }
+        Consumer<T> listenerOnResponseAssert = (r) -> {
+            assert r.getFailure() instanceof ReplicationOperation.RetryOnPrimaryException == false :
+                "IndexShard shouldn't use RetryOnPrimaryException. got " + r.getFailure();
+            listener.onResponse(r);
+        };
         if (result.getResultType() == Engine.Result.Type.MAPPING_UPDATE_REQUIRED) {
+            Mapping mappingUpdate = result.getRequiredMappingUpdate();
             try {
-                mappingUpdate.updateMappings(result.getRequiredMappingUpdate(), shardId, Constants.DEFAULT_MAPPING_TYPE);
+                validateMapping(mappingUpdate.root().iterator(), false);
             } catch (Exception e) {
-                return onMappingUpdateError.apply(e);
-            }
-            result = execute.get();
-            if (result.getResultType() == Engine.Result.Type.MAPPING_UPDATE_REQUIRED) {
-                // double mapping update. We assume that the successful mapping update wasn't yet processed on the node
-                // and retry the entire request again.
-                throw new ReplicationOperation.RetryOnPrimaryException(shardId,
-                    "Dynamic mappings are not available on the node that holds the primary yet");
+                listener.onResponse(onMappingUpdateError.apply(e));
+                return;
             }
+            schemaUpdateClient.updateOnMaster(
+                shardId.getIndex(),
+                mappingUpdate,
+                ActionListener.wrap(
+                    ack -> {
+                        T r;
+                        try {
+                            r = execute.get();
+                        } catch (Exception e) {
+                            listener.onFailure(e);
+                            return;
+                        }
+                        if (r.getResultType() == Engine.Result.Type.MAPPING_UPDATE_REQUIRED) {
+                            // double mapping update. We assume that the successful mapping update
+                            // wasn't yet processed on the node and retry the entire request again.
+                            listener.onFailure(new ReplicationOperation.RetryOnPrimaryException(
+                                shardId, "Dynamic mappings are not available on the node that holds the primary yet"));
+                        } else {
+                            listenerOnResponseAssert.accept(r);
+                        }
+                    },
+                    e -> listener.onResponse(onMappingUpdateError.apply(e))
```

Comment:
also here I think this should be `listener.onFailure()`

Reply:
No, the original code:

```try {
                mappingUpdate.updateMappings(result.getRequiredMappingUpdate(), shardId, Constants.DEFAULT_MAPPING_TYPE);
            } catch (Exception e) {
                return onMappingUpdateError.apply(e);
            }```